### ISSUE TITLE
Lazy Dataset loaders

### ISSUE THREAD

**Issue by TristanBilot** [TristanBilot/mlx-graphs#14]
Dataset loaders in PyG loads all the data into memory, then returns batches of graphs for computation. However, all the remaining graphs not used for computation are still present in memory even if not used in the current batch.
This may be not important on traditional computers with separate CPU and GPU memory, as we load the entire dataset in CPU memory and perform the expensive computations on the batches separately on GPU.

In mlx, CPU and GPU memory is the same with unified memory. This means that loading a large dataset reduces the available memory for computations on GPU. 

Using a lazy `Dataset` or `DataLoader` that only yields batches one by one would prevent from loading the entire dataset and then batching.

**Comment by francescofarina** [comment]
Yes, that's interesting. Is that death with in any ways in `mlx-data`?

**Comment by francescofarina** [comment]
I was looking at `mlx-data` in more details. I think we could just define a standard `Data` in `mlx_graphs` as an object with a prescribed structure (e.g., `node_features`, `edge_index`, ...) 
```
@dataclass
class Data:
    node_features: mx.array
    edge_index: mx.array
    ...
```
and then use it directly with `mlx-data` by converting it to a dict
```
import mlx.data as dx

_dataset = [Data(...), ..., Data(...)]

dataset = (
    dx.buffer_from_vector(_dataset)
    .shuffle()
    .to_stream()
    ...
)
```

I'll give it a go tomorrow

**Comment by francescofarina** [comment]
So, after the discussion in https://github.com/TristanBilot/mlx-graphs/pull/16 I think that we need two classes of datasets:
- one will basically be treated as a list of `Data`, all living in memory
- one which will likely mimic the [`OnDiskDataset`](https://pytorch-geometric.readthedocs.io/en/latest/_modules/torch_geometric/data/on_disk_dataset.html#OnDiskDataset) in PyG

The second one will likely be useful only for very large datasets.
